File: src/shared/api/imageLibrary.js
import { invoke } from '@tauri-apps/api/core'
import { convertFileSrc } from '@tauri-apps/api/core'

async function uint8ArrayToNumberArrayChunked(data, chunkSize = 256 * 1024) {
  if (!data) return []
  if (Array.isArray(data)) return data
  const u8 = data instanceof Uint8Array ? data : new Uint8Array(data)
  const result = new Array(u8.length)
  for (let offset = 0; offset < u8.length; offset += chunkSize) {
    const end = Math.min(u8.length, offset + chunkSize)
    for (let i = offset; i < end; i++) {
      result[i] = u8[i]
    }
    await new Promise(resolve => setTimeout(resolve, 0))
  }
  return result
}

// 初始化图片库目录
export async function initImageLibrary() {
  return await invoke('il_init')
}

// 保存图片
export async function saveImage(group, filename, data) {
  const payloadData = await uint8ArrayToNumberArrayChunked(data)
  return await invoke('il_save_image', {
    payload: { group, filename, data: payloadData }
  })
}

// 获取图片列表
export async function getImageList(group, offset = 0, limit = 20) {
  return await invoke('il_get_image_list', {
    payload: { group, offset, limit }
  })
}

// 获取图片总数
export async function getImageCount(group) {
  return await invoke('il_get_image_count', {
    payload: { group }
  })
}

// 删除图片
export async function deleteImage(group, filename) {
  return await invoke('il_delete_image', {
    payload: { group, filename }
  })
}

// 重命名图片
export async function renameImage(group, oldFilename, newFilename) {
  return await invoke('il_rename_image', {
    payload: { group, old_filename: oldFilename, new_filename: newFilename }
  })
}

// 获取图库分组
export async function getImageGroups() {
  return await invoke('il_get_groups')
}

// 新增图库分组
export async function addImageGroup(name, icon = 'ti ti-photo', color = '#2563eb') {
  return await invoke('il_add_group', {
    payload: { name, icon, color }
  })
}

// 更新图库分组
export async function updateImageGroup(oldName, newName, icon = 'ti ti-photo', color = '#2563eb') {
  return await invoke('il_update_group', {
    payload: { old_name: oldName, new_name: newName, icon, color }
  })
}

// 移动图片到分组
export async function moveImageToGroup(sourceGroup, filename, targetGroup) {
  return await invoke('il_move_image_to_group', {
    payload: { source_group: sourceGroup, filename, target_group: targetGroup }
  })
}

// 删除图库分组
export async function deleteImageGroup(name, moveImagesToDefault = false) {
  return await invoke('il_delete_group', {
    payload: { name, move_images_to_default: moveImagesToDefault }
  })
}

// 获取图片目录路径
export async function getImagesDir() {
  return await invoke('il_get_images_dir')
}

// 获取 GIF 目录路径
export async function getGifsDir() {
  return await invoke('il_get_gifs_dir')
}

// 将本地路径转换为URL
export function getImageUrl(path) {
  return convertFileSrc(path)
}

File: src-tauri/src/commands/image_library.rs
use serde::Deserialize;
use crate::services::image_library;
use std::time::Duration;

#[derive(Deserialize)]
pub struct SaveImagePayload {
    group: String,
    filename: String,
    data: Vec<u8>,
}

#[derive(Deserialize)]
pub struct GetImageListPayload {
    group: String,
    offset: usize,
    limit: usize,
}

#[derive(Deserialize)]
pub struct GetImageCountPayload {
    group: String,
}

#[derive(Deserialize)]
pub struct DeleteImagePayload {
    group: String,
    filename: String,
}

#[derive(Deserialize)]
pub struct RenameImagePayload {
    group: String,
    old_filename: String,
    new_filename: String,
}

#[derive(Deserialize)]
pub struct ImageGroupPayload {
    name: String,
    icon: String,
    color: String,
}

#[derive(Deserialize)]
pub struct RenameImageGroupPayload {
    old_name: String,
    new_name: String,
    icon: String,
    color: String,
}

#[derive(Deserialize)]
pub struct MoveImagePayload {
    source_group: String,
    filename: String,
    target_group: String,
}

#[derive(Deserialize)]
pub struct DeleteImageGroupPayload {
    name: String,
    move_images_to_default: bool,
}

#[tauri::command]
pub fn il_init() -> Result<(), String> {
    image_library::init_image_library()
}

#[tauri::command]
pub async fn il_save_image(payload: SaveImagePayload) -> Result<image_library::ImageInfo, String> {
    let group = payload.group;
    let filename = payload.filename;
    let data = payload.data;

    let handle = tokio::task::spawn_blocking(move || image_library::save_image(&group, &filename, &data));
    match tokio::time::timeout(Duration::from_secs(15), handle).await {
        Ok(join_result) => join_result.map_err(|e| format!("任务执行失败: {}", e))?,
        Err(_) => Err("保存图片超时".to_string()),
    }
}

#[tauri::command]
pub fn il_get_image_list(payload: GetImageListPayload) -> Result<image_library::ImageListResult, String> {
    image_library::get_image_list(&payload.group, payload.offset, payload.limit)
}

#[tauri::command]
pub fn il_get_image_count(payload: GetImageCountPayload) -> Result<usize, String> {
    image_library::get_image_count(&payload.group)
}

#[tauri::command]
pub fn il_delete_image(payload: DeleteImagePayload) -> Result<(), String> {
    image_library::delete_image(&payload.group, &payload.filename)
}

#[tauri::command]
pub fn il_rename_image(payload: RenameImagePayload) -> Result<image_library::ImageInfo, String> {
    image_library::rename_image(&payload.group, &payload.old_filename, &payload.new_filename)
}

#[tauri::command]
pub fn il_get_images_dir() -> Result<String, String> {
    let path = image_library::get_images_dir()?;
    Ok(path.to_string_lossy().to_string())
}

#[tauri::command]
pub fn il_get_gifs_dir() -> Result<String, String> {
    let path = image_library::get_gifs_dir()?;
    Ok(path.to_string_lossy().to_string())
}

#[tauri::command]
pub fn il_get_groups() -> Result<Vec<image_library::ImageGroupInfo>, String> {
    image_library::list_groups()
}

#[tauri::command]
pub fn il_add_group(payload: ImageGroupPayload) -> Result<image_library::ImageGroupInfo, String> {
    image_library::add_group(&payload.name, &payload.icon, &payload.color)
}

#[tauri::command]
pub fn il_update_group(payload: RenameImageGroupPayload) -> Result<image_library::ImageGroupInfo, String> {
    image_library::update_group(&payload.old_name, &payload.new_name, &payload.icon, &payload.color)
}

#[tauri::command]
pub fn il_move_image_to_group(payload: MoveImagePayload) -> Result<image_library::ImageInfo, String> {
    image_library::move_image_to_group(&payload.source_group, &payload.filename, &payload.target_group)
}

#[tauri::command]
pub fn il_delete_group(payload: DeleteImageGroupPayload) -> Result<Vec<image_library::ImageGroupInfo>, String> {
    image_library::delete_group(&payload.name, payload.move_images_to_default)
}
